pkg/crypto/rings/cache.go

```diff
@@ -0,0 +1,201 @@
+package rings
+
+import (
+	"context"
+	"log"
+	"sync"
+
+	"cosmossdk.io/depinject"
+	ring_secp256k1 "github.com/athanorlabs/go-dleq/secp256k1"
+	ringtypes "github.com/athanorlabs/go-dleq/types"
+	cosmosclient "github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	accounttypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/noot/ring-go"
+
+	"github.com/pokt-network/poktroll/pkg/relayer"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+)
+
+var _ RingCache = (*ringCache)(nil)
+
+type ringCache struct {
+	// ringPointsCache maintains a map of application addresses to the points
+	// on the secp256k1 curve that correspond to the public keys of the gateways
+	// the application is delegated to. These are used to build the app's ring.
+	ringPointsCache map[string][]ringtypes.Point
+	ringPointsMu    *sync.RWMutex
+
+	// clientCtx is the client context for the application, and is used to query
+	// the application and account modules.
+	clientCtx relayer.QueryClientContext
+
+	// applicationQuerier is the querier for the application module, and is
+	// used to get the addresses of the gateways an application is delegated to.
+	applicationQuerier apptypes.QueryClient
+
+	// accountQuerier is the querier for the account module, and is used to get
+	// the public keys of the application and its delegated gateways.
+	accountQuerier accounttypes.QueryClient
+}
+
+// NewRingCache returns a new RingCache instance. It requires a depinject.Config
+// to be passed in, which is used to inject the dependencies of the RingCache.
+func NewRingCache(deps depinject.Config) (RingCache, error) {
+	rc := &ringCache{
+		ringPointsCache: make(map[string][]ringtypes.Point),
+		ringPointsMu:    &sync.RWMutex{},
+	}
+
+	if err := depinject.Inject(
+		deps,
+		&rc.clientCtx,
+	); err != nil {
+		return nil, err
+	}
+
+	clientCtx := cosmosclient.Context(rc.clientCtx)
+
+	rc.accountQuerier = accounttypes.NewQueryClient(clientCtx)
+	rc.applicationQuerier = apptypes.NewQueryClient(clientCtx)
+
+	return rc, nil
+}
+
+// GetRingForAddress returns the ring for the address provided. If it does not
```

Comment:
>  If it does not exist in the cache, it will be created by querying the application module

Congrats, you've rediscovered memorization: https://en.wikipedia.org/wiki/Memoization

Reply:
I actually did this one in class recently lmao (recursion and backtracking) 🤓